File: q0931/q0931_iterative.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {

        // Let dp[i, j] represents the min path sum until position (i, j)
        int[][] dp = new int[matrix.length][matrix[0].length];
        for (int j = 0; j < matrix[0].length; j++)
            dp[0][j] = matrix[0][j];

        // To arrive at position (i, j), there are three different ways
        for (int i = 1; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                // Either from directly up, or up left, or up right excluding the invalid
                // positions), use placeholder value Integer.MAX_VALUE to remove this
                // possibility during the Math.min step
                int up = dp[i - 1][j];
                int upLeft = j > 0 ? dp[i - 1][j - 1] : Integer.MAX_VALUE;
                int upRight = j < matrix[0].length - 1 ? dp[i - 1][j + 1] : Integer.MAX_VALUE;
                dp[i][j] = matrix[i][j] + Math.min(up, Math.min(upLeft, upRight));
            }
        }

        // Final result is the max value of the last row of the dp
        int pathSum = Integer.MAX_VALUE;
        for (int j = 0; j < matrix[0].length; j++) {
            pathSum = Math.min(pathSum, dp[matrix.length - 1][j]);
        }
        return pathSum;
    }
}

File: q0931/q0931_iterative.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int res = INT_MAX;
        int n = matrix.size();
        vector<vector<int>> dp(n, vector<int>(n, 0));

        for (int r = n - 1; r >= 0; --r) {
            for (int c = 0; c < n; ++c) {
                if (r == n - 1) {
                    dp[r][c] = matrix[r][c];
                } else if (c == 0) {
                    dp[r][c] = matrix[r][c] + min(dp[r + 1][c], dp[r + 1][c + 1]);
                } else if (c == n - 1) {
                    dp[r][c] = matrix[r][c] + min(dp[r + 1][c], dp[r + 1][c - 1]);
                } else {
                    dp[r][c] = matrix[r][c] + min({dp[r + 1][c - 1], dp[r + 1][c], dp[r + 1][c + 1]});
                }
            }
        }

        // Taking the minimum of the first row
        for (int i = 0; i < n; ++i) {
            res = min(res, dp[0][i]);
        }

        return res;
    }
};